python/tvm/contrib/debugger/debug_runtime.py

```diff
@@ -0,0 +1,541 @@
+            #del node['inputs']
+            node['inputs'] = input_list
+            dltype = str("type: " + dltype_list[1][i])
+            if 'attrs' not in node:
+                node['attrs'] = {}
+                node['op'] = "param"
+            else:
+                node['op'] = node['attrs']['func_name']
+            node['name'] = node['name'].replace("/", "_")
+            node['attrs'].update({"T": dltype})
+            node['shape'] = shapes_list[1][i]
+            p_graph['nodes'].append(node)
+        return p_graph
+
+    def _get_debug_buffer_count(self):
+        return len(self.dbg_buff_list)
+
+    def _get_debug_buffer(self, eid):
+        return self.dbg_buff_list[eid]
+
+    def set_debug_buffer(self):
+        """Set output buffer allocated for each node to copy the node's
+        output after Run completed.
+
+        This function will get called before run performs.
+        GraphRuntime copy the execution out to the allocated memory for each nodes.
+
+        Parameters
+        ----------
+        None
+        """
+        for eid in range(self._get_debug_buffer_count()):
+            self._set_debug_buffer(self._get_debug_buffer(eid))
+
+    def _create_debug_ui(self, p_graph, nodes_list, heads_list, ctx, frontend):
+        """Create UI wrapper framework to handle multiple UI frontends for tvmdbg
+
+        Parameters
+        ----------
+        p_graph : json format
+            json formatted NNVM graph contain list of each node's
+            name, shape and type.
+
+        nodes_list : list
+            List of all the nodes presented in the graph
+
+        heads_list : list
+            List of all output nodes presented in the graph
+
+        ctx : TVMContext
+            The context this module is under.
+
+        frontend: str
+            'cli'- involve curses based CLI frontend
+            'tensorboard'- make data format for tensorbard frontend.
+
+        Returns
+        -------
+        ui_wrapper : DebugGraphUIWrapper object
+            UI warpper manage tvmdbg frontend.
+        """
+        ctx = str(ctx).upper().replace("(", ":").replace(")", "")
+        ui_wrapper = DebugGraphUIWrapper(p_graph, nodes_list,
+                                         heads_list, ctx, frontend)
+        return ui_wrapper
+
+    def _make_debug_buffer_list(self, shapes_list, dltype_list):
+        """Allocate output buffer for each node to copy the node's
+        output after Run completed.
+
+        Parameters
+        ----------
+        shapes_list: list
+            List of shape of each nodes presented in the graph.
+
+        dltype_list: list
+            List of data type of each nodes presented in the graph.
+
+        Returns
+        -------
+        dbg_out_buffer_list : list
+            Allocated empty buffer
+        """
+        dbg_out_buffer_list = []
+        for i in range(len(shapes_list[1])):
+            dbg_out_buffer_list.append(nd.empty(shapes_list[1][i], dltype_list[1][i]))
```

Comment:
just to make sure I understand correctly: these buffers always stay on cpu, which are for storing the network outputs for debug.

Reply:
yes. always stay in cpu. for storing the outputs of each fused op after execution, 
This will be updated after each op execution.